### PR TITLE
V2 Edit Venue journey

### PR DESCRIPTION
**Pull request by gunndabad**
V2-erizes the exiting Edit Venue journey and adds functionality for editing website, phone number and email.

Before:
image

After:
image

### CODE REVIEW

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+        [Fact]
+        public async Task Get_ValidRequestNoExistingFormFlowInstance_RendersExpectedOutputFromDatabase()
+        {
+            // Arrange
+            var providerId = await TestData.CreateProvider();
+            var venueId = await TestData.CreateVenue(
+                providerId,
+                venueName: "Test Venue",
+                email: "test-venue@provider.com",
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs:65]: I shy away from using real domain names. Suggest using `example.com` which is reserved.

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                telephone: "01234 567890",
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs:66]: This is a good source of test phone numbers that don't risk being anyone's real number https://www.ofcom.org.uk/phones-telecoms-and-internet/information-for-industry/numbering/numbers-for-drama

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/SaveTests.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        [InlineData(TestUserType.ProviderSuperUser)]
+        [InlineData(TestUserType.ProviderUser)]
+        public async Task Post_UserCannotAccessVenue_ReturnsForbidden(TestUserType userType)
+        {
+            // Arrange
+            var providerId = await TestData.CreateProvider(ukprn: 12345);
+            var venueId = await TestData.CreateVenue(providerId, email: "person@provider.com");
+
+            var anotherProviderId = await TestData.CreateProvider(ukprn: 67890);
+
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/SaveTests.cs:34]: Whitespace error here according to git show. Remove trailing whitespace. We should update editorconfig to trim trailing spaces.

File: src/Dfc.CourseDirectory.Web/Controllers/VenuesController.cs [hunk 4]
```diff
@@ -223,38 +234,7 @@ public async Task<IActionResult> AddVenueManualAddress(string Id)
 
             return View(viewModel);
         }
-        [Authorize]
-        public async Task<IActionResult> EditVenue(string Id)
```
Comment (timabell) [inline src/Dfc.CourseDirectory.Web/Controllers/VenuesController.cs:227]: I think this should be replaced with a redirect to the new url in case any users have bookmarked etc.

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            var request = new HttpRequestMessage(HttpMethod.Get, $"venues/{venueId}");
+
+            // Act
+            var response = await HttpClient.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            var doc = await response.GetDocument();
+            Assert.Equal("Updated name", doc.GetSummaryListValueWithKey("Location name"));
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs:123]: I prefer to use a shared `const string` for the test values as it makes it clear to the reader and compiler that the values for setup and assertion are both required to be the same.

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            Assert.Equal(
+                "Updated line 1\nUpdated line 2\nUpdated town\nUpdated county\nUP1 D8D",
+                doc.GetSummaryListValueWithKey("Address"));
+            Assert.Equal("updated@provider.com", doc.GetSummaryListValueWithKey("Email"));
+            Assert.Equal("02345 678901", doc.GetSummaryListValueWithKey("Phone"));
+            Assert.Equal("updated-provider.com", doc.GetSummaryListValueWithKey("Website"));
```
Comment (timabell) [inline tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs:129]: I'd switch all the assertions to FluentAssertions as we've started to do elsewhere for better error messages and slightly nicer readability of test code.
I'd also wrap these blocks of assertions in an `AssertionScope` so that if there are problems you get all the errors at once. https://fluentassertions.com/introduction#assertion-scopes

**Review by timabell — COMMENTED** [review]
Great work, very good to see this all moving to a more testable and better designed system.

Observations:

* The v2 code is very abstract, and would not be comprehensible to 99% of the c# devs I've met if you weren't there to explain it. The patterns and libraries in use needs documenting somewhere in the repo that the team can point to by way of explanation. Without those this PR is hard to follow. Areas I noticed that could use explanation:
	* Use of mediator.
	* Use of CQRS patterns in `Features`.
	* Concept and use of FormFlow.
	* Validation system.
	* Use of `OneOf`/`Match` library.
	* Caching of entities.
* I suggest adding some bogart links that inspired the designs to the readme under a new section "## Code Architecture" too.
* The tests seem quite repetitive, lots of similar Forbidden/NotFound tests etc, can we DRY this? I'm worried that when we want to change things the sheer volume of test code will slow us down.
* I think the onspd commits could be squashed together (facd8860 and b68f7de9), they seem like a single iteration to me.
* Put the PTCD number at start of the first line of the commit message for dee26485 "v2 venue details page", so that we can grep it when doing release notes.
* I think we'd agreed to standardize on having the PTCD at the start of the commit message rather than the end in brackets as this PR currently has.

With the comments on the code, it's assumed that where I've commented on one occurence that any changes will be applied to all occurrences in the PR.

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/EmailTests.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            var request = new HttpRequestMessage(HttpMethod.Get, $"venues/{venueId}/email");
+
+            // Act
+            var response = await HttpClient.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            var doc = await response.GetDocument();
+            Assert.Equal("person@provider.com", doc.GetElementById("Email").GetAttribute("value"));
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/EmailTests.cs:74]: Generally I agree with using test IDs but in this case I'm not so sure; these element IDs mirror the model property names. I'm not sure what using test IDs buys in this case.

[the remark this answers is not in the document]
Comment (timabell) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/EmailTests.cs:74]: If you are reading just the markup it draws your attention to the fact that there is a test that depends on this markup, so you know to maintain that properly when making changes.

File: src/Dfc.CourseDirectory.Core/Validation/VenueValidation/RuleBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        private static readonly Regex _addressLinePattern = new Regex(
+            @"^[a-zA-Z0-9\.\-']+(?: [a-zA-Z0-9\.\-']+)*$",
+            RegexOptions.Compiled);
+
+        private static readonly Regex _countyPattern = new Regex(
+            @"^[a-zA-Z\.\-']+(?: [a-zA-Z\.\-']+)*$",
+            RegexOptions.Compiled);
+
+        private static readonly Regex _townPattern = new Regex(
+            @"^[a-zA-Z\.\-']+(?: [a-zA-Z\.\-']+)*$",
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Core/Validation/VenueValidation/RuleBuilderExtensions.cs:21]: This is copied old UI to keep existing behaviour

File: src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessVenueBehavior.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                _providerContextProvider.SetProviderContext(new ProviderContext(providerInfo));
+
+                return await next();
+            }
+            else
+            {
+                throw new NotAuthorizedException();
+            }
+
+            bool IsAuthorized()
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessVenueBehavior.cs]: Done

File: src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessVenueBehavior.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            var VenueId = _descriptor.GetVenueId(request);
+            var providerId = await _providerOwnershipCache.GetProviderForVenue(VenueId);
+
+            if (providerId == null)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.Venue, VenueId);
+            }
+
+            var isAuthorized = IsAuthorized();
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessVenueBehavior.cs]: Done

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs (hunk 5 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs:123]: > * The tests seem quite repetitive, lots of similar Forbidden/NotFound tests etc, can we DRY this? I'm worried that when we want to change things the sheer volume of test code will slow us down.

I've not found a simple way of doing this. From doing this style of tests for a number of years I haven't yet hit issues with the repition.

**Comment by gunndabad** [comment]
> * The tests seem quite repetitive, lots of similar Forbidden/NotFound tests etc, can we DRY this? I'm worried that when we want to change things the sheer volume of test code will slow us down.

I've not found a simple way of doing this. From doing this style of tests for a number of years I haven't yet hit issues with the repition.

File: src/Dfc.CourseDirectory.Core/Validation/VenueValidation/RuleBuilderExtensions.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (timabell) [reply src/Dfc.CourseDirectory.Core/Validation/VenueValidation/RuleBuilderExtensions.cs:21]: Now that is a perfectly good reason. Move along, nothing to see here.

**Comment by timabell** [comment]
As mentioned on slack I had a go at a bit of DRY to see what shape it would take, and ended up with this rough draft https://github.com/SkillsFundingAgency/dfc-coursedirectory/compare/venue-additions...venue-additions-dry2

On the subject of why I think the repetition is problematic see https://twitter.com/tim_abell/status/1311994247197925378

File: src/Dfc.CourseDirectory.WebV2/Features/EditVenue/EditVenueFlowModel.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                Name = venue.VenueName,
+                PhoneNumber = venue.Telephone,
+                Website = venue.Website,
+                AddressLine1 = venue.AddressLine1,
+                AddressLine2 = venue.AddressLine2,
+                Town = venue.Town,
+                County = venue.County,
+                Postcode = venue.Postcode,
+                Latitude = venue.Latitude,
+                Longitude = venue.Longitude
```
[the remark this answers is not in the document]
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Features/EditVenue/EditVenueFlowModel.cs:57]: I see the argument here but I don't agree - it's aesthetically jarring to me

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs (hunk 2 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs]: Done

File: src/Dfc.CourseDirectory.Web/Controllers/VenuesController.cs (hunk 4 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Web/Controllers/VenuesController.cs:227]: Done

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/SaveTests.cs (hunk 3 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/SaveTests.cs]: Fixed

File: tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs (hunk 6 above)
Comment (gunndabad) [reply tests/Dfc.CourseDirectory.WebV2.Tests/FeatureTests/EditVenue/DetailsTests.cs]: Done

**Review by soharadesign — APPROVED** [review]

**Review by timabell — APPROVED** [review]
:boom: shipit
:ship: :ship: :ship: :ship:
